File: src/Atlasd/Battlenet/Protocols/HTTP/HttpListener.cs
using Atlasd.Daemon;
using System;
using System.Collections.Generic;
using System.Net;
using System.Net.Sockets;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace Atlasd.Battlenet.Protocols.Http
{
    class HttpListener : IDisposable, IListener
    {
        public bool IsListening { get => Socket != null; } /* part of IListener */
        public IPEndPoint LocalEndPoint { get; protected set; } /* part of IListener */
        public Socket Socket { get; protected set; } /* part of IListener */

        public HttpListener(IPEndPoint endp)
        {
            SetLocalEndPoint(endp);
        }

        public void Close() /* part of IListener */
        {
            Stop();
        }

        public void Dispose() /* part of IDisposable */
        {
            Close();
        }

        private void ProcessAccept(SocketAsyncEventArgs e)
        {
            if (e.SocketError != SocketError.Success)
            {
                Logging.WriteLine(Logging.LogLevel.Error, Logging.LogType.Http, e.RemoteEndPoint, "HTTP listener socket error occurred!");
            }

            // Start the read loop on a new stack
            Task.Run(new HttpSession(e.AcceptSocket).ConnectedEvent);

            // Accept the next connection request
            StartAccept(e);
        }

        public void SetLocalEndPoint(IPEndPoint endp) /* part of IListener */
        {
            if (IsListening)
            {
                throw new InvalidOperationException();
            }

            LocalEndPoint = endp;
        }

        public void Start() /* part of IListener */
        {
            Stop();

            Logging.WriteLine(Logging.LogLevel.Info, Logging.LogType.Server, $"Starting HTTP listener on [{LocalEndPoint}]");

            Socket = new Socket(LocalEndPoint.AddressFamily, SocketType.Stream, System.Net.Sockets.ProtocolType.Tcp)
            {
                ExclusiveAddressUse = true,
                NoDelay = Daemon.Common.TcpNoDelay,
                UseOnlyOverlappedIO = true,
            };
            Socket.Bind(LocalEndPoint);
            Socket.Listen(-1);

            StartAccept(null);
        }

        private void StartAccept(SocketAsyncEventArgs acceptEventArg)
        {
            if (acceptEventArg == null)
            {
                acceptEventArg = new SocketAsyncEventArgs();
                acceptEventArg.Completed += new EventHandler<SocketAsyncEventArgs>(AcceptEventArg_Completed);
            }
            else
            {
                // socket must be cleared since the context object is being reused
                acceptEventArg.AcceptSocket = null;
            }

            bool willRaiseEvent = Socket.AcceptAsync(acceptEventArg);
            if (!willRaiseEvent)
            {
                ProcessAccept(acceptEventArg);
            }
        }

        void AcceptEventArg_Completed(object sender, SocketAsyncEventArgs e)
        {
            ProcessAccept(e);
        }

        public void Stop() /* part of IListener */
        {
            if (!IsListening) return;

            Logging.WriteLine(Logging.LogLevel.Info, Logging.LogType.Server, $"Stopping HTTP listener on [{Socket.LocalEndPoint}]");

            Socket.Close();
            Socket = null;
        }
    }
}

File: src/AtlasV/Battlenet/Protocols/HTTP/HttpListener.vb
Imports AtlasV.Daemon
Imports System
Imports System.Collections.Generic
Imports System.Net
Imports System.Net.Sockets
Imports System.Text
Imports System.Threading
Imports System.Threading.Tasks

Namespace AtlasV.Battlenet.Protocols.Http
    Public Class HttpListener
        Implements IDisposable, IListener

        Public ReadOnly Property IsListening As Boolean Implements IListener.IsListening
            Get
                Return Socket IsNot Nothing
            End Get
        End Property

        Public Property LocalEndPoint As IPEndPoint Implements IListener.LocalEndPoint
        Public Property Socket As Socket Implements IListener.Socket

        Public Sub New(ByVal varEndPoint As IPEndPoint)
            SetLocalEndPoint(varEndPoint)
        End Sub

        Public Sub Close() Implements IListener.Close
            [Stop]()
        End Sub

        Public Sub Dispose() Implements IDisposable.Dispose
            Close()
        End Sub

        Private Sub ProcessAccept(ByVal varSocketAsyncEventArgs As SocketAsyncEventArgs)
            If varSocketAsyncEventArgs.SocketError <> SocketError.Success Then
                Logging.WriteLine(Logging.LogLevel.[Error], Logging.LogType.Http, varSocketAsyncEventArgs.RemoteEndPoint, "HTTP listener socket error occurred!")
            End If

            Task.Run(Sub()
                         Dim tmpSession As New HttpSession(varSocketAsyncEventArgs.AcceptSocket)
                         tmpSession.ConnectedEvent()
                     End Sub).Wait()
            'Task.Run(New HttpSession(varSocketAsyncEventArgs.AcceptSocket).ConnectedEvent()) 'how does this even work on c#, this line makes no sense.
            StartAccept(varSocketAsyncEventArgs)
        End Sub

        Public Sub SetLocalEndPoint(ByVal varEndPoint As IPEndPoint) Implements IListener.SetLocalEndPoint
            If IsListening Then
                Throw New InvalidOperationException()
            End If

            LocalEndPoint = varEndPoint
        End Sub

        Public Sub Start() Implements IListener.Start
            [Stop]()
            Logging.WriteLine(Logging.LogLevel.Info, Logging.LogType.Server, $"Starting HTTP listener on [{LocalEndPoint}]")
            Socket = New Socket(LocalEndPoint.AddressFamily, SocketType.Stream, System.Net.Sockets.ProtocolType.Tcp) With {
                .ExclusiveAddressUse = True,
                .NoDelay = Daemon.Common.TcpNoDelay,
                .UseOnlyOverlappedIO = True
            }
            Socket.Bind(LocalEndPoint)
            Socket.Listen(-1)
            StartAccept(Nothing)
        End Sub

        Private Sub StartAccept(ByVal varAcceptEventArg As SocketAsyncEventArgs)
            If varAcceptEventArg Is Nothing Then
                varAcceptEventArg = New SocketAsyncEventArgs()
                AddHandler varAcceptEventArg.Completed, New EventHandler(Of SocketAsyncEventArgs)(AddressOf AcceptEventArg_Completed)
            Else
                varAcceptEventArg.AcceptSocket = Nothing
            End If

            Dim willRaiseEvent As Boolean = Socket.AcceptAsync(varAcceptEventArg)

            If Not willRaiseEvent Then
                ProcessAccept(varAcceptEventArg)
            End If
        End Sub

        Private Sub AcceptEventArg_Completed(ByVal sender As Object, ByVal varSocketAsyncEventArgs As SocketAsyncEventArgs)
            ProcessAccept(varSocketAsyncEventArgs)
        End Sub

        Public Sub [Stop]() Implements IListener.Stop
            If Not IsListening Then Return
            Logging.WriteLine(Logging.LogLevel.Info, Logging.LogType.Server, $"Stopping HTTP listener on [{Socket.LocalEndPoint}]")
            Socket.Close()
            Socket = Nothing
        End Sub

    End Class
End Namespace